src/mono/mono/mini/interp/interp.c

```diff
@@ -3945,6 +3945,36 @@ mono_interp_exec_method (InterpFrame *frame, ThreadContext *context, FrameClause
 						gpointer unboxed = mono_object_unbox_internal (this_arg);
 						LOCAL_VAR (0, gpointer) = unboxed;
 					}
+
+					InterpMethodCodeType code_type = new_method->code_type;
+
+					g_assert (code_type == IMETHOD_CODE_UNKNOWN ||
+							code_type == IMETHOD_CODE_INTERP ||
+							code_type == IMETHOD_CODE_COMPILED);
+
+					if (G_UNLIKELY (code_type == IMETHOD_CODE_UNKNOWN)) {
+						// FIXME push/pop LMF
+						MonoMethodSignature *sig = mono_method_signature_internal (new_method->method);
+						if (mono_interp_jit_call_supported (new_method->method, sig))
+							code_type = IMETHOD_CODE_COMPILED;
+						else
+							code_type = IMETHOD_CODE_INTERP;
+						new_method->code_type = code_type;
+					}
+
+					if (code_type == IMETHOD_CODE_COMPILED) {
+						frame->imethod = new_method;
+						ip = frame->imethod->code;
+						frame->state.ip = ip;
```

Comment:
The code here is somewhat dubious to me. Since we determined that the method that we wanted to tailcall is not interp method, then we can no longer do a tailcall. So we shouldn't overwrite the contents of the calling frame. It should be like we just do a jitcall and then exit the frame. The memmove above shouldn't happen if we do a jitcall and we shouldn't really update `frame->imethod`, and `frame->state.ip`. It probably doesn't make a difference in practice, but the code might make more sense.

Reply:
Fixed.